plugins/generic/externalFeed/controllers/grid/ExternalFeedGridHandler.inc.php

```diff
@@ -0,0 +1,244 @@
+		}
+		
+	}
+
+	//
+	// Overridden methods from GridHandler
+	//
+	/**
+	 * @copydoc Gridhandler::getRowInstance()
+	 */
+	function getRowInstance() {
+		return new ExternalFeedGridRow();
+	}
+
+	//
+	// Public Grid Actions
+	//
+	/**
+	 * Display the grid's containing page.
+	 * @param $args array
+	 * @param $request PKPRequest
+	 */
+	function index($args, $request) {
+		$templateMgr = TemplateManager::getManager($request);
+		$dispatcher = $request->getDispatcher();
+		return $templateMgr->fetchAjax(
+			'externalFeedGridContainer',
+			$dispatcher->url(
+				$request, ROUTE_COMPONENT, null,
+				'plugins.generic.externalFeed.controllers.grid.ExternalFeedGridHandler',
+				'fetchGrid'
+			)
+		);
+	}
+
+	/**
+	 * An action to add a new external feed
+	 * @param $args array Arguments to the request
+	 * @param $request PKPRequest Request object
+	 */
+	function addExternalFeed($args, $request) {
+		return $this->editExternalFeed($args, $request);
+	}
+
+	/**
+	 * An action to edit an external feed
+	 * @param $args array Arguments to the request
+	 * @param $request PKPRequest Request object
+	 * @return string Serialized JSON object
+	 */
+	function editExternalFeed($args, $request) {
+		$feedId = $request->getUserVar('feedId');
+		$context = $request->getContext();
+		$this->setupTemplate($request);
+		// Create and present the edit form
+		import('plugins.generic.externalFeed.controllers.grid.form.ExternalFeedForm');
+		$externalFeedPlugin = self::$plugin;
+		$externalFeedForm = new ExternalFeedForm(self::$plugin, $context->getId(), $feedId);
+		$externalFeedForm->initData();
+		return new JSONMessage(true, $externalFeedForm->fetch($request));
+	}
+
+	/**
+	 * Update feed
+	 * @param $args array
+	 * @param $request PKPRequest
+	 * @return string Serialized JSON object
+	 */
+	function updateExternalFeed($args, $request) {
+		$feedId = $request->getUserVar('feedId');
+		$context = $request->getContext();
+		$this->setupTemplate($request);
+	
+		import('plugins.generic.externalFeed.controllers.grid.form.ExternalFeedForm');
+		$externalFeedPlugin = self::$plugin;
+		$externalFeedForm = new ExternalFeedForm(self::$plugin, $context->getId(), $feedId);
+		$externalFeedForm->readInputData($request);
+	
+		if ($externalFeedForm->validate()) {
+			$externalFeedForm->execute();
+			return DAO::getDataChangedEvent();
+		} else {
+			// Present any errors
+			$json = new JSONMessage(true, $externalFeedForm->fetch($request));
+			return $json->getString();
+		}
+	}
+
+	/**
+	 * Delete an external feed entry
+	 * @param $args array
+	 * @param $request PKPRequest
+	 * @return string Serialized JSON object
+	 */
+	function delete($args, $request) {
+		$feedId = $request->getUserVar('feedId');
+		$context = $request->getContext();
+		
+		$externalFeedDao = DAORegistry::getDAO('ExternalFeedDAO');
+		if ($externalFeedDao->getExternalFeedJournalId($feedId) == $context->getId()) {
+			$externalFeedDao->deleteExternalFeedById($feedId);

```

Comment:
By adding a `contextId` parameter to this function, we can make integrity checks inherent rather than checking that the context ID matches first and then deleting conditionally. There should never be a `delete` without a context check so it's good to build it in. With that done, and my other comment about code calling `deleteExternalFeedById` (which as I've also noted should be renamed to `deleteById`), it should be possible to remove `getExternalFeedJournalId` entirely.

Reply:
@asmecher : if my understanding is correct, `contextId` should be added to `deleteById` signature so we would have `function deleteById($feedId, $contextId)`, right? 

How about `ExternalFeedDAO::deleteObject` which relies on `deleteById`.  Should `deleteObject` also receive `contextId` as parameter? I am asking because I looked around and none of DAO::deleteObject seem to do so.

Reply:
That's right (with the addition of the optional second parameter) -- it should be

    function deleteById($feedId, $contextId = null)

The `deleteObject` function is OK not to specify the second parameter to `deleteById`, as these objects can only be instantiated safely, i.e. in ways where the context ID will be the appropriate value.

Reply:
Okay. thanks.